aten/src/ATen/native/mps/operations/HistogramKernel.mm

```diff
@@ -0,0 +1,395 @@
+#include <ATen/native/Histogram.h>
+#include <ATen/native/mps/OperationUtils.h>
+
+namespace at::native {
+namespace mps {
+
+enum BIN_SELECTION_ALGORITHM {
+  LINEAR_INTERPOLATION,
+  LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH,
+  BINARY_SEARCH,
+};
+
+static const char* METAL_HISTOGRAM = R"HISTOGRAM_METAL(
+
+#include <metal_stdlib>
+using namespace metal;
+
+enum BIN_SELECTION_ALGORITHM {
+  LINEAR_INTERPOLATION,
+  LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH,
+  BINARY_SEARCH,
+};
+
+// Re-implementation of std::upper_bound with some modifications.
+template<typename T>
+int64_t upper_bound(constant T * arr, int64_t first, int64_t len, T val) {
+  int64_t middle;
+  int64_t half_;
```

Comment:
Even though int64 is supported by MPS, I think there are pretty fundamental limitation on accessing tensors larger than 2**32 elements, so should we use `int32_t` instead (or make it a template argument) 
Another question: why not use an unsigned type here?

Reply:
For the reason not using unsigned type: https://github.com/pytorch/pytorch/pull/96652#discussion_r1152841014

I can make it a template.